### PR TITLE
save tile entity packets too

### PR DESCRIPTION
**Pull request by u9g**
I also made gamemode creative default because it's helpful for getting the blocks that you need like signs quickly for testing

### CODE REVIEW

**Comment by rom1504** [comment]
please rebase

**Review by rom1504 — CHANGES_REQUESTED** [review]
this module test coverage is pretty much 100%
I actually did TDD here

Let's keep it like this, can you adapt tests ?

**Comment by u9g** [comment]
> this module test coverage is pretty much 100%
> I actually did TDD here
> 
> Let's keep it like this, can you adapt tests ?

I 

> this module test coverage is pretty much 100%
> I actually did TDD here
> 
> Let's keep it like this, can you adapt tests ?

I can't adapt tests as there are very few naturally spawning block entities, the only way I could think of is if I make a mineflayer bot during continuous saving to use a block entity

**Comment by u9g** [comment]
> please rebase

done

**Comment by rom1504** [comment]
the tests check the existence of files, not their content
you can def do that

**Comment by u9g** [comment]
> the tests check the existence of files, not their content
> you can def do that

def do what, what if there is no block entities?

**Comment by rom1504** [comment]
if there are no block entities there should still be an empty file

but anyway, if there are no block entities, how do you intend to use this ?

**Comment by u9g** [comment]
> if there are no block entities there should still be an empty file
> 
> but anyway, if there are no block entities, how do you intend to use this ?

It doesnt work like that, it puts a file when the packet happens with the location of the block entity. You use it with continuous capture, by going into the world and doing it yourself (ie: place a sign, use a cauldron, etc)

**Comment by rom1504** [comment]
yeah that doesn't make sense with the design of this package
could you please adapt the code so it's more useful ?

https://github.com/PrismarineJS/minecraft-chunk-dumper#usage is the normal use

in particular it's used to dump 10 chunks for a version and add that to pchunk test folder https://github.com/PrismarineJS/prismarine-chunk/tree/master/test

**Comment by u9g** [comment]
> yeah that doesn't make sense with the design of this package
> 
> could you please adapt the code so it's more useful ?
> 
> 
> 
> https://github.com/PrismarineJS/minecraft-chunk-dumper#usage is the normal use
> 
> 
> 
> in particular it's used to dump 10 chunks for a version and add that to pchunk test folder https://github.com/PrismarineJS/prismarine-chunk/tree/master/test

so how do you suggest I make the block entities get there?

**Comment by rom1504** [comment]
same as the rest of the package. You wait for block entities chunks to be as big as what the user asked

**Comment by rom1504** [comment]
the end result regardless of the way should be we have N chunks with light, chunk and block entities files
otherwise it won't be possible to use that in pchunk tests

**Comment by u9g** [comment]
> same as the rest of the package. You wait for block entities chunks to be as big as what the user asked

its not a chunk though, its 1 block that pops up, so do I just keep getting chunks until I collect x number if blocks?

**Comment by rom1504** [comment]
yes

**Comment by rom1504** [comment]
ah and yeah you're using the wrong packet
tile_entity_data is useful only for old versions

for new versions, block entities are part of the chunk packet

**Comment by u9g** [comment]
> ah and yeah you're using the wrong packet
> 
> tile_entity_data is useful only for old versions
> 
> 
> 
> for new versions, block entities are part of the chunk packet

so this pr isnt needed since chunk packet is already dumped?

**Comment by rom1504** [comment]
no this part of the packet is not dumped currently
and the old packet is still needed for old versions

**Comment by u9g** [comment]
> no this part of the packet is not dumped currently
> 
> and the old packet is still needed for old versions

what is the old packet, and for what minecraft versions is it?

**Comment by rom1504** [comment]
tile_entity_data is the old packet

it's used until 1.15 iirc

**Comment by u9g** [comment]
thats false, I got this from 1.16.5

**Comment by rom1504** [comment]
https://wiki.vg/Protocol#Chunk_Data

tile_entity_data is the update packet, similar to block update one

**Comment by u9g** [comment]
> https://wiki.vg/Protocol#Chunk_Data
> 
> 
> 
> tile_entity_data is the update packet, similar to block update one

Oh, so instead I should just be fetching that. So again, is this not needed? Just manually add to the chunk packet to unpack block entities

**Comment by rom1504** [comment]
yes it's needed
https://github.com/PrismarineJS/minecraft-chunk-dumper/blob/62aaf0eb983fb51285d5e9e07b366a772440170e/index.js#L39

**Comment by rom1504** [comment]
would be great to finish

**Comment by rom1504** [comment]
can you use this to produce data and put it in https://github.com/PrismarineJS/prismarine-chunk/pull/113 ?
last check to make sure it works as expected

File: index.js [hunk 1]
```diff
@@ -184,10 +183,12 @@ class ChunkDumper extends EventEmitter {
...
         // always have a block entity to record
-        this.server.writeServer(`/op ${this.client.username}\n`)
-        setTimeout(() => {
-          this.client.write('chat', { message: '/setblock ~ ~ ~1 beacon' })
-        }, 100)
+        this.client.on('login', () => {
+          this.server.writeServer(`/op ${this.client.username}\n`)
+          setTimeout(() => {
+            this.client.write('chat', { message: '/setblock ~ ~ ~1 beacon' })
+          }, 100)
```
Comment (rom1504) [inline index.js:189]: 1000 no ?

File: index.js [hunk 2]
```diff
@@ -184,10 +183,12 @@ class ChunkDumper extends EventEmitter {
         }
         this.on('tile_entity', saveTileEntities)
         // always have a block entity to record
-        this.server.writeServer(`/op ${this.client.username}\n`)
-        setTimeout(() => {
-          this.client.write('chat', { message: '/setblock ~ ~ ~1 beacon' })
-        }, 100)
+        this.client.on('login', () => {
```
Comment (rom1504) [inline index.js]: login prob not enough, just wait 1000

**Comment by rom1504** [comment]
merged the other one, please fix conflicts

**Pull request closed without merging**